wolfcrypt/src/chacha20_poly1305.c

```diff
@@ -376,27 +389,32 @@ int wc_ChaCha20Poly1305_Final(ChaChaPoly_Aead* aead,
     byte outAuthTag[CHACHA20_POLY1305_AEAD_AUTHTAG_SIZE])
 {
     int ret = 0;
-    byte padding[CHACHA20_POLY1305_MAC_PADDING_ALIGNMENT - 1];
-    word32 paddingLen;
-    byte little64[16]; /* word64 * 2 */
 
     if (aead == NULL || outAuthTag == NULL) {
         return BAD_FUNC_ARG;
     }
+
+    /* make sure AAD is calculated */
+    ret = wc_ChaCha20Poly1305_CalcAad(aead);
+
     if (aead->state != CHACHA20_POLY1305_STATE_DATA) {
```

Comment:
'if (aead->state != CHACHA20_POLY1305_STATE_DATA)' will never be true since aead==NULL has already been checked and wc_ChaCha20Poly1305_CalcAad() will always set it.  I think you can remove the three lines.

Reply:
I do not agree.
1. `wc_ChaCha20Poly1305_CalcAad` will only set `CHACHA20_POLY1305_STATE_DATA` if state is `CHACHA20_POLY1305_STATE_AAD`. This check is still required.
2. Yes there is a duplicate "aead == NULL" check in `wc_ChaCha20Poly1305_CalcAad`, but that is acceptable for future protection in case that function is called from new functions or is ever made public...

Reply:
@dgarske, you are correct. I missed the CHACHA20_POLY1305_STATE_AAD check before advancing the state. I guess it would catch a misuse of doing wc_ChaCha20Poly1305_Init() and wc_ChaCha20Poly1305_Final() without an *_update() call in between.

Optionally, I suspect a check would be faster than a function call, you should consider using
if (aead->state == CHACHA20_POLY1305_STATE_AAD) 
          ret = wc_ChaCha20Poly1305_CalcAad(aead);

I think you also need to advance the state in wc_ChaCha20Poly1305_UpdateData() with 
        aead->state = CHACHA20_POLY1305_STATE_DATA;
Otherwise, Init/Update/Final of data only without AAD won't work.

Reply:
Thanks Tesfa. Excellent feedback. Updates pushed to https://github.com/wolfSSL/wolfssl/commit/784d95afbec81827e5a6dd5380cc171a64e5d9ce and customer tests revalidated.